### DISCUSSION TITLE
Revert peer deps to npm 6

### DISCUSSION THREAD

**Discussion by wesbos** [npm/feedback#610] [CLI] [answered]
Has this been discussed at all? Peer dependencies are such a pain to deal with.

It's been over a year with npm 7, and there are no signs of the ecosystem getting better at this.

Every time a major version of a library or framework comes out, all the plugins (most of which work fine without any issue) are unable to be installed until the author can update the peer deps.  By the time the ecosystem has caught up, there is another major version of that library and we start all over again.

It's at a point now where I have to say in my tutorials "Just run it with --legacy-peer-deps everytime" and big frameworks like Gatsby have it in the docs as just something you need to do. 

If we need to use that flag to install anything, should it be reverted to the default and just warn people?  I feel like it's doing more harm than good

**Comment by alcuadrado** [comment]
Not to be the grinch, but I have a completely different take.

I'm a maintainer of Hardhat, the most popular Ethereum development tool, which has a plugin system. We rely very heavily on peer dependencies and really benefited from npm 7. Installing plugins became trivial, and now we can have some level of certainty that the user didn't mess up their versions.

The key difference may be that we are extremely cautious about releasing major versions, even more so of the "core" package. This isn't something we changed because of npm 7, but decided at the beginning of the project. We aim to only release a major version every 1.5-2 years, and the only time we did it, we coordinated the release with the authors of all the major community plugins.

I agree that peer dependencies are still pretty painful, but I don't think this change would be in the right direction. To me not checking their compatibility at installation time means hoping for the system to work and making the user responsible of dealing with the problems when it doesn't.

**Reply by wesbos** [reply]
Yeah this is totally fair and a good example of peer deps working as they should, but I still feel that for every one story of a package being released every 1.5-2 years, we have React, Gatsby, Next.js, ESLint and hundreds of other plugins that release a few times a year. 

I guess the problem here is that:

*  peer dependencies aren't usually a good way to find out if a package works with the next major version. With most libs in the JS space, they warn on 1 version, and remove the API on the next.
*  libraries aren't updated in time for major versions

I don't know what the answer is here, but I do know it's currently very painful for almost everyone I talk to. I remember years ago WordPress had a user thumbs up/down for plugins to solicit feedback from the community.

**Comment by ljharb** [comment]
All these problems still existed in npm < 7, you just didn’t know about them.

Any app where `npm ls` exits nonzero has an invalid dep graph, and you can’t rely on anything working. This means, if a package screws up peer deps, that package is *broken* and you shouldn’t use it at all.

npm 7+ doesn’t go far enough imo - it surfaces these issues on install, which is a massive improvement, but it actually lets you proceed when things are invalid. I’d prefer the entire app to be uninstallable unless the dep graph is valid :-)

**Reply by wesbos** [reply]
If that were the case I'd get literally zero work done. I've don't have a single production app that can be installed without --legacy-peer-deps.

Like, I agree that it should work that way, but it doesn't. What am I supposed to do to get my work done if a package that 100% works, is missing a peer dep for the version of a library that came out 4 days ago?

I don't really understand how it worked in npm 6, but I do know everything worked great for me

**Reply by ljharb** [reply]
It does it you take the time to clean up your app’s dependency graph, which I’ve done at many of the companies I’ve worked for.

**Reply by wesbos** [reply]
:\ "dont use that package" is the solution here?

**Reply by travi** [reply]
> a library that came out 4 days ago?

usually, patience has been the answer for me. sure, i often cannot upgrade to a brand new base library, but many of the most popular packages take care of updating their peer dependency definitions fairly quickly after such things happen. there are stragglers, but i've usually been successful getting those to accept a PR that adds the new version to the peer dependency definition.

**Reply by ljharb** [reply]
@wesbos yes, just like if it has a bug in it. Broken peer deps is a bug, and why would anyone choose to use buggy software, or to upgrade to a buggy version?

**Reply by wesbos** [reply]
It's not a bug though. I get what you are saying and I wish it was as perfect as whatever you work on, just in my experience it's always a huge pain for regular devs

**Reply by ljharb** [reply]
It absolutely is a bug. It’s just that npm has tolerated it for so long that developers haven’t internalized this truth.

npm 7+’s stricter behavior will change this attitude, eventually, and this temporary annoyance will be largely replaced by ecosystem correctness.

**Reply by wesbos** [reply]
I sure hope so, that is what they said a year ago. How much longer do you expect this to be true for?

**Reply by ljharb** [reply]
npm < 7 was out for 9 years, but i suspect much faster than that. Does it matter how long? The future is longer than the past, and npm will be relevant much longer than the time it takes for this to improve.

**Reply by arcanis** [reply]
> It absolutely is a bug. It’s just that npm has tolerated it for so long that developers haven’t internalized this truth.

It's not, though. The peer dependency contract is that the instance matches the host, and the version check is a hint but not a mandatory requirement. I know you believe otherwise and I won't change *your* mind, but you should probably be less assertive - npm did change semantic. Sure, they did it for reason they thought were right, but they did change it nonetheless. They can take responsibility for the aftermath.

**Reply by travi** [reply]
i can understand how it can been seen as not-a-bug when things happen to continue working even though the peer-dependency range suggests incompatibility. 

however, consider it from the other direction. peer-dependencies are a package maintainer's way to tell the tools, and therefore users, when certain versions are _known to be incompatible_. if this information is ignored by the tools, the first sign that things are broken are through errors in your application at runtime. you were told that it would be broken, but that warning was ignored. 

defining this correctly does add additional work on maintainers and causes trouble when packages aren't maintained actively, but i agree that the tool feedback is the best way to bring attention to the current lax approach that many have taken when defining peer-dependencies and hopefully improve how maintainers approach what they define.

**Reply by ljharb** [reply]
It’d always been required because that’s how npm ls treated it.

what npm changed is that they closed a loophole where some folks *got away with* treating peer deps as not required.

**Reply by arcanis** [reply]
It's still a mystery to me why you use `npm ls` as a reference for whether an install is valid instead of, say, *the install command*.

But that's not my software, I don't mind if that's how you see things, just don't say it's an universal truth 🤷‍♀️

**Reply by travi** [reply]
> that is what they said a year ago. How much longer do you expect this to be true for?

at that point, it was more opt-in to use npm v7. the earliest version of node that bundled it directly was node v16, which only recently became LTS. i expect many more folks will become aware of this change soon and likely give more and more feedback to maintainers where peer-dependencies are defined in an incompatible way.

hopefully there are some that will help steer the conversation in the direction of correcting the peer-dependency definitions rather than only suggesting use of the `--legacy-peer-deps` flag.

**Reply by ljharb** [reply]
@arcanis there's not "an install", there's only "a dep graph" (a package.json, maybe a lockfile, and node_modules) and `npm ls`'s semantics were always to check its validity. "installing" just creates a dep graph that may or may not be valid (which is still the case)

**Reply by markgoho** [reply]
> the earliest version of node that bundled it directly was node v16, which only recently became LTS. i expect many more folks will become aware of this change soon and likely give more and more feedback to maintainers where peer-dependencies are defined in an incompatible way.

this 👆🏻

I think most folks are still on node 12 or 14 because most lib maintainers are only _now_ deprecating 10, it's going to be years before a critical mass gets to 16 (node 22 will be out then) and we'll start hearing more about this

**Reply by wesbos** [reply]
Yeah thats a good point, hopefully that puts the pressure on!

**Comment by MylesBorins** [comment, marked as answer]
Hey @wesbos. At this time reverting the peer deps change from npm 7 / 8 is not something that we are actively considering. We are currently planning for npm 9 and if you would like to seriously pursue this the next course of action would be to open an RFC in https://github.com/npm/rfcs to suggest the change.
